### PR TITLE
Support of read\write access separation for "find usages"

### PR DESCRIPTION
**Pull request by mayakwd**

### CODE REVIEW

File: src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.intellij.plugins.haxe.lang.psi.*;
+import com.intellij.plugins.haxe.lang.util.HaxePsiUtil;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiReference;
+import org.jetbrains.annotations.NotNull;
+
+public class HaxeReadWriteAccessDetector extends ReadWriteAccessDetector {
+  @Override
+  public boolean isReadWriteAccessible(@NotNull final PsiElement element) {
+    return element instanceof HaxePsiField || element instanceof HaxeClass || element instanceof HaxeMethod;
```
[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java]: It's definitely will for local variables, but about array access I'm not sure

File: src/common/com/intellij/plugins/haxe/lang/util/HaxePsiUtil.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/lang/util/HaxePsiUtil.java]: How about we rename this file HaxeExpressionUtil?  We already have a number of PsiUtil classes and they are used for inspecting and manipulating the PSI.  All of these functions are about expressions anyway.

File: src/common/com/intellij/plugins/haxe/lang/util/HaxePsiUtil.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  public static boolean isIncrementDecrementOperation(@Nullable PsiElement element) {
+    if (element instanceof HaxePrefixExpression) {
+      final IElementType operationType = HaxePsiUtil.getOperationType((HaxePrefixExpression)element);
+      return HaxeTokenTypeSets.UNARY_READ_WRITE_OPERATORS.contains(operationType);
+    }
+    else return element instanceof HaxeSuffixExpression;
+  }
+
+  @Nullable
+  public static PsiElement skipParenthesizedExprUp(@Nullable PsiElement parent) {
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/lang/util/HaxePsiUtil.java]: This is the only function that is PSI-specific, and it can be private because it's only used in this class.  Otherwise, you might actually want it in UsefulPsiTreeUtil.

File: src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (EricBishton) [reply src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java]: The trouble with array is: Even if the declaration states that the array itself is not write accessible, it's elements _are_.

[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java]: `arr[0] = 12;` is recognizing as read operation
`arr = [];` as write operation

It works in the same way as it java does. Should we make it recognize `arr[0] = 12;` as write operation?

[the remark this answers is not in the document]
Comment (RealyUniqueName) [reply src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java]: I think you're right. No need to make it a write operation.

[the remark this answers is not in the document]
Comment (EricBishton) [reply src/common/com/intellij/plugins/haxe/ide/HaxeReadWriteAccessDetector.java]: Is there a place for both?  I get that arr[0] *is* really just an array access.  It's also nice to be able to see that the array _contents_ are being modified.  But then again, if somebody takes a copy of the entry and modifies it, it would be very difficult for us to track that change, too.
 ```Haxe
  var myobj = arr[0];
  myobj.setSomeValue(0);  // Hard to follow that back to the read and mark that as read/write.
```
